src/analysis/gMCS/GPR2models.m

```diff
@@ -0,0 +1,259 @@
+% This function look into the GPRparser and builds auxiliary nodes for each
+% layer of the GPR rules.
+% For example:
+%     g1 & g2 & g3 is one layer in the model,   RXN = g1 + g2 + g3
+%     (g1 | g2) & g3 is a two layer model:      RXN = (g1 | g2) + g3
+%                                               (g1 | g2) = g1
+%                                               (g1 | g2) = g2   
+
+
+
+%  Name nodes according to position and layer inside the GPR rule
+node = cell(size(head.children));
+for child=1:length(head.children)
+    node{child} = [node_parent,'_',num2str(child)];
+end
+model = addMetabolite(model, node);
+
+%  Expand model with every node, using OR/AND rules
+if strcmp(class(head),'OrNode')
+    for child=1:length(head.children)
+        model = addReaction(model,node{child},...
+            'reactionFormula',[node{child},' -> ',parent],'printLevel',printLevel-1);
+    end
+else
+    model = addReaction(model,strjoin(node,' & '),...
+        'reactionFormula',[strjoin(node,' + '),' -> ',parent],'printLevel',printLevel-1);
+end
+
+
+for i=1:length(head.children)
+    child = head.children(i);
+    % check if the child is final layer
+    if strcmp(class(child),'LiteralNode')
+        gen = genes(str2num(child.id));
+        model = addReaction(model,[gen{1},'_',node{i}],...
+            'metaboliteList',{gen{1},node{i}},'stoichCoeffList',[-1, 1],...
+            'printLevel',printLevel-1);
+    else
+        model = modelParser(model,genes,child,node{i},node{i},printLevel);
+    end
+end
+
+end
+
+function modelOut = reduceModel(model)
+% As the algorithm generates redundant nodes, this function is used to
+% remove these nodes:
+% Ej:       gen -> node -> rule  == gen -> rule
+
+% This is done for every gene of the model
+genes = model.genes;
+for gen=1:length(genes)
+    gen_idx = find(strcmp(model.mets,genes{gen})); % find idx of gene
+    rxn_gen_out = find(model.S(gen_idx,:)<0);   % find reactions that consume gene
+    % for each reaction, find consumption node
+    [node_idx,aux] = find(model.S(:,rxn_gen_out)>0);
+    for i = 1:length(node_idx)
+        % for each reaction, find consumption node
+        node_name = model.mets(node_idx(i));
+        % find reactions out of this node
+        [aux, rxn_idx] = find((model.S(node_idx(i),:)<0));
+        rxn_name = model.rxns(rxn_idx);
+        % in this reaction, change (node -> rule) for (gene -> rule)
+        model = changeRxnMets(model, node_name, genes(gen), rxn_name);   
+    end
+    % remove reactions (gene -> node)
+%     model = removeRxns(model, model.rxns(rxn_gen_out), 'metFlag',false); 
+    % remove stoichimetri of intermediate nodes
+    model.S(node_idx,:)=0;
+    % remove output of genes to intermediate nodes
+    model.S(:,rxn_gen_out)=0;
+end
+
+% remove unused reactions and unused metabolites
+% model = removeTrivialStoichiometry(model);    % problems, removes mets 2 times
+anyRowBool = any(model.S,2);
```

Comment:
Could you use removeFieldEntriesforType here?
This would ensure, that no model fields (also in the future) are missed during the resizing.